File: 863-all-nodes-distance-k-in-binary-tree/863-all-nodes-distance-k-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def distanceK(self, root: TreeNode, target: TreeNode, k: int) -> List[int]:
        parentMap = {root: None}
        queue = [root]
        
        while queue:
            node = queue.pop(0)
            if node.left != None:
                queue.append(node.left)
                parentMap[node.left] = node
            if node.right != None:
                queue.append(node.right)
                parentMap[node.right] = node
        
        visited = {target: True}
        queue = [(target, 0)]
        ans = []
        while queue:
            node, dist = queue.pop(0)
            if dist == k:
                ans.append(node.val)
            if node.left != None:
                if node.left not in visited:
                    queue.append((node.left, dist+1))
                    visited[node.left] = True
            if node.right != None:
                if node.right not in visited:
                    queue.append((node.right, dist+1))
                    visited[node.right] = True
            if node != root and parentMap[node] not in visited:
                queue.append((parentMap[node], dist+1))
                visited[parentMap[node]] = True
            
        return ans
            
            

File: 863-all-nodes-distance-k-in-binary-tree/863-all-nodes-distance-k-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        Map<TreeNode, TreeNode> parentMap = new HashMap<>();
        Map<TreeNode, Boolean> visited = new HashMap<>();
        List<TreeNode> queue = new ArrayList<>();
        queue.add(root);
        while(queue.size() != 0) {
            TreeNode node = queue.remove(0);
            if(node.left != null) {
                queue.add(node.left);
                parentMap.put(node.left, node);
            }
            if(node.right != null) {
                queue.add(node.right);
                parentMap.put(node.right, node);
            }
        }
        
        int dist = 0;
        List<TreeNode> newqueue = new ArrayList<>();
        newqueue.add(target);
        visited.put(target, true);
        while(!newqueue.isEmpty()) {
            int size = newqueue.size();
            if(dist == k)
                break;
            
            dist++;
            for(int i=0; i< size; i++) {
                TreeNode node = newqueue.remove(0);
                if(node.left != null && visited.get(node.left) == null) {
                    newqueue.add(node.left);
                    visited.put(node.left, true);
                }
                if(node.right != null && visited.get(node.right) == null) {
                    newqueue.add(node.right);
                    visited.put(node.right, true);
                }
                if(parentMap.get(node) != null && visited.get(parentMap.get(node)) == null) {
                    newqueue.add(parentMap.get(node));
                    visited.put(parentMap.get(node), true);
                }                    
            }
        }
        List<Integer> res = new ArrayList<>();
        for(TreeNode node: newqueue)
            res.add(node.val);
        
        return res;
    }
}